### PR TITLE
Defer the server shutdown until the clients finish calls to avoid tim…

### PR DESCRIPTION
**Pull request by veblush**
This is to fix the flakiness of  `//test/cpp/qps:json_run_localhost_*` which happened to have timeout. Culprit is that the client couldn't detect that a server was just shut down and the call that client sent to the server was stuck without a proper error propagation. This only happens 1 or less in 10000 runs. To be resilient to this problem, this test is modified to have a graceful test shutdown process.

Internal issue: b/153650457

### CODE REVIEW

**Comment by veblush** [comment]
Huh, `//test/cpp/qps:json_run_localhost_cpp_protobuf_async_streaming_from_server_qps_unconstrained` doesn't seem to work with this change.

File: test/cpp/qps/driver.cc [hunk 1]
```diff
@@ -541,6 +530,16 @@ std::unique_ptr<ScenarioResult> RunScenario(
     rrc->set_count(it->second);
   }
 
+  gpr_log(GPR_INFO, "Finishing servers");
```
Comment (jtattermusch) [inline test/cpp/qps/driver.cc:533]: can you please elaborate a bit more on why this change?
Because the flake only happens pretty rarely, we need to be really sure that the change will have the desired effect.

Comment (veblush) [reply test/cpp/qps/driver.cc]: The main reason why it's timed out is that the client is stuck by sending the last request to the server which has been already shutdown. To address this, this PR let client finish first to give it a chance to clean up everything while server is live and then let server finish.

File: test/cpp/qps/driver.cc [hunk 2]
```diff
@@ -492,6 +492,10 @@ std::unique_ptr<ScenarioResult> RunScenario(
   Histogram merged_latencies;
   std::unordered_map<int, int64_t> merged_statuses;
 
+  // For the case where clients leads the test, it's going to finish
+  // clients first and wait until it's completely done for generaous termination.
```
Comment (vjpai) [inline test/cpp/qps/driver.cc:496]: generaous ? Do you mean graceful ?

File: test/cpp/qps/driver.cc [hunk 3]
```diff
@@ -504,16 +508,19 @@ std::unique_ptr<ScenarioResult> RunScenario(
...
-    if (!server->stream->Write(server_mark)) {
-      gpr_log(GPR_ERROR, "Couldn't write mark to server %zu", i);
-      GPR_ASSERT(false);
-    }
-    if (!server->stream->WritesDone()) {
-      gpr_log(GPR_ERROR, "Failed WritesDone for server %zu", i);
-      GPR_ASSERT(false);
+
+  if (!client_finish_first) {
+    gpr_log(GPR_INFO, "Finishing servers");
```
Comment (vjpai) [inline test/cpp/qps/driver.cc:513]: There's so much common content here between the two cases; do you want to move this to a side function?

File: test/cpp/qps/driver.cc [hunk 4]
```diff
@@ -492,6 +492,10 @@ std::unique_ptr<ScenarioResult> RunScenario(
   Histogram merged_latencies;
   std::unordered_map<int, int64_t> merged_statuses;
 
+  // For the case where clients leads the test, it's going to finish
+  // clients first and wait until it's completely done for generaous termination.
+  bool client_finish_first = (client_config.rpc_type() != STREAMING_FROM_SERVER);
+
```
Comment (vjpai) [inline test/cpp/qps/driver.cc:498]: Can you document why this is the case? I'm not sure that I understand the difference here (it's been a while since I saw the code)

File: test/cpp/qps/driver.cc (hunk 2 above)
Comment (veblush) [reply test/cpp/qps/driver.cc:496]: Oops. Right!

File: test/cpp/qps/driver.cc (hunk 3 above)
Comment (veblush) [reply test/cpp/qps/driver.cc:513]: Sure I'm going to refactor this large function into many small functions.

**Comment by veblush** [comment]
Thanks, Vijay. I confirmed that it passes all following tests with 10000x.

```
--config=dbg //test/cpp/qps:json_run_localhost_cpp_protobuf_sync_unary_ping_pong_insecure@poller=epoll1
--config=opt //test/cpp/qps:json_run_localhost_cpp_generic_async_streaming_qps_unconstrained_64KBmsg_insecure@poller=epollex
--config=opt //test/cpp/qps:json_run_localhost_cpp_protobuf_async_client_sync_server_unary_qps_unconstrained_insecure@poller=epollex
--config=tsan //test/cpp/qps:json_run_localhost_cpp_protobuf_async_unary_qps_unconstrained_2waysharedcq_secure@poller=epoll1
--config=tsan //test/cpp/qps:json_run_localhost_cpp_protobuf_sync_unary_ping_pong_secure@poller=epollex
--config=tsan //test/cpp/qps:json_run_localhost_cpp_protobuf_sync_unary_ping_pong_secure@poller=poll
--config=ubsan //test/cpp/qps:json_run_localhost_cpp_protobuf_sync_unary_ping_pong_insecure@poller=epollex
--config=ubsan //test/cpp/qps:json_run_localhost_cpp_protobuf_sync_unary_ping_pong_secure@poller=poll
```
